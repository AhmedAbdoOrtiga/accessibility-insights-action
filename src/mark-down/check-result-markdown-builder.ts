// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { injectable } from 'inversify';

import { brand } from '../content/strings';
import { bold, footerSeparator, heading, link, listItem, productTitle, sectionSeparator } from '../utils/markdown-formatter';
import { CombinedReportParameters } from 'accessibility-insights-report';

@injectable()
export class CheckResultMarkdownBuilder {
    public errorContent = (): string => {
        const lines = [
            heading(`${productTitle()}: Something went wrong`, 3),
            sectionSeparator(),

            `You can review the log to troubleshoot the issue. Fix it and re-run the workflow to run the automated accessibility checks again.`,
            sectionSeparator(),
        ];

        return this.scanResultDetails(lines.join(''));
    };

    public content = (combinedReportParameters: CombinedReportParameters): string => {
        const passedChecks = combinedReportParameters.results.resultsByRule.passed.length;
        const inapplicableChecks = combinedReportParameters.results.resultsByRule.notApplicable.length;
        const failedChecks = combinedReportParameters.results.resultsByRule.failed.reduce((a, b) => a + b.failed.length, 0);

        const lines = [
            heading(`${productTitle()}`, 3),
            // sectionSeparator(),
            listItem(`${bold(`Rules: `)}${failedChecks} check(s) failed, ${passedChecks} check(s) passed, and ${inapplicableChecks} were not applicable`),
            // sectionSeparator(),
            // sectionSeparator(),
            listItem(
                `${bold(`URLs: `)}${combinedReportParameters.results.urlResults.failedUrls} URL(s) failed, ${combinedReportParameters.results.urlResults.passedUrls} URL(s) passed, and ${combinedReportParameters.results.urlResults.unscannableUrls} were not scannable`,
            ),
            sectionSeparator(),

            this.downloadArtifacts(),
        ];

        return this.scanResultDetails(lines.join(''), this.scanResultFooter(combinedReportParameters));
    };

    private readonly scanResultDetails = (scanResult: string, footer?: string): string => {
        const lines = [scanResult, sectionSeparator(), footerSeparator(), sectionSeparator(), footer];

        return lines.join('');
    };

    private readonly scanResultFooter = (combinedReportParameters: CombinedReportParameters): string => {
        const axeVersion = combinedReportParameters.axeVersion;
        const axeCoreUrl = `https://github.com/dequelabs/axe-core/releases/tag/v${axeVersion}`;
        const axeLink = link(axeCoreUrl, `axe-core ${axeVersion}`);

        return `This scan used ${axeLink} with ${combinedReportParameters.userAgent}.`;
    };



    private readonly downloadArtifacts = () => {
        return listItem(`Download the ${bold(brand)} artifact to view the detailed results of these checks`);
    };
}
