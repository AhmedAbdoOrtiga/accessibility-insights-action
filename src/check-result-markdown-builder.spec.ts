// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
// tslint:disable:no-import-side-effect no-any
import 'reflect-metadata';

import { AxeScanResults } from 'accessibility-insights-scan';

import { CheckResultMarkdownBuilder } from './check-result-markdown-builder';

// tslint:disable: no-unsafe-any no-null-keyword no-object-literal-type-assertion
describe(CheckResultMarkdownBuilder, () => {
    let checkResultMarkdownBuilder: CheckResultMarkdownBuilder;
    let axeScanResults: AxeScanResults;

    beforeEach(() => {
        checkResultMarkdownBuilder = new CheckResultMarkdownBuilder();
    });

    it('getFailureDetails', () => {
        axeScanResults = {
            results: {
                testEngine: {
                    version: '1.0.1',
                },
                violations: [
                    {
                        id: 'color-contrast',
                        nodes: [{ html: 'html1' }, { html: 'html2' }],
                        description: 'color-contrast is important',
                    },
                    {
                        id: 'duplicate-id',
                        nodes: [{ html: 'html3' }, { html: 'html4' }, { html: 'html5' }],
                        description: 'duplicate-id is even more important',
                    },
                ],
                passes: [{ html: 'passed' }],
                inapplicable: [{ html: 'inapplicable' }],
            },
            browserSpec: 'browserSpec',
        } as any;

        expect(checkResultMarkdownBuilder.getFailureDetails(axeScanResults)).toMatchSnapshot();
    });

    it('errorContent', () => {
        expect(checkResultMarkdownBuilder.errorContent()).toMatchSnapshot();
    });

    it('congratsContent', () => {
        axeScanResults = {
            browserSpec: 'browserSpec',
            results: {
                testEngine: {
                    version: '1.0.1',
                },
                violations: [],
                passes: [{ html: 'passed' }],
                inapplicable: [{ html: 'inapplicable' }],
            },
        } as any;

        const res = checkResultMarkdownBuilder.congratsContent(axeScanResults);

        expect(res).toMatchSnapshot();
    });
});
