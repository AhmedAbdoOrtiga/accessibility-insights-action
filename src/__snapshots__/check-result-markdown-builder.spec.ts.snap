// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckResultMarkdownBuilder congratsContent 1`] = `
"## Accessibility Automated Checks Results
The Accessibility Insights Service ran a set of automated checks to help find some of the most common accessibility issues. The best way to evaluate web accessibility compliance is to complete a [WCAG 2.1 compliance assessment](http://aka.ms/AccessibilityInsights).
###### DETAILS
### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights: All applicable checks passed
* **1 check(s) passed**, and 1 were not applicable
* Download the **Accessibility Insights** artifact to view the detailed results of these checks.
---
This scan used [axe-core 1.0.1](https://github.com/dequelabs/axe-core/releases/tag/v1.0.1) with the following browsers: browserSpec."
`;

exports[`CheckResultMarkdownBuilder errorContent 1`] = `
"## Accessibility Automated Checks Results
The Accessibility Insights Service ran a set of automated checks to help find some of the most common accessibility issues. The best way to evaluate web accessibility compliance is to complete a [WCAG 2.1 compliance assessment](http://aka.ms/AccessibilityInsights).
###### DETAILS
### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights: Something went wrong
You can review the log to troubleshoot the issue. Fix it and re-run the workflow to run the automated accessibility checks again.

---
"
`;

exports[`CheckResultMarkdownBuilder getFailureDetails 1`] = `
"## Accessibility Automated Checks Results
The Accessibility Insights Service ran a set of automated checks to help find some of the most common accessibility issues. The best way to evaluate web accessibility compliance is to complete a [WCAG 2.1 compliance assessment](http://aka.ms/AccessibilityInsights).
###### DETAILS
### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights
* **2 check(s) failed**, 1 passed, and 1 were not applicable
* Download the **Accessibility Insights** artifact to view the detailed results of these checks.
#### Failed instances
* **2 × color-contrast**:  color-contrast is important
* **3 × duplicate-id**:  duplicate-id is even more important

---
This scan used [axe-core 1.0.1](https://github.com/dequelabs/axe-core/releases/tag/v1.0.1) with the following browsers: browserSpec."
`;
